### PR TITLE
WIP: GridPositionMultiStoreUpdater implementation for CMS Pages

### LINKED ISSUE #18965: BO - Pages CMS - Position column should be displayed only in a specific store context when the multistore is enabled
#### Is your feature request related to a problem?

We should have the same behavior on all BO grid in multistore when there is a position column

#### Describe the solution you'd like

In Design > Pages, the position column should be displayed only in a specific store context when the multistore is enabled (same behavior as on Categories page) and I should be able to move pages only in a specific store context

### PR DESCRIPTION
**Pull request by kpodemski**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Possibility to sort entities in relation to shop table
| Type?         | improvement
| Category?     | CO
| BC breaks?    | yes
| Deprecations? | no
| Fixed ticket? | Fixes #18965
| How to test?  | Play with CMS pages drag&drop sort, with and without categories, with & without multi-store

## Some thoughts

1. Is it good to almost copy `DoctrinePositionUpdateHandler `? I didn't want to introduce to much changes to core for this, we have an easy way (IFs), and a bit more clear solution introduced with this PR **but** it is something to think about.
2. Implementation in `CategoryGrid` is cleaner but I think that it would be better for developers to have it (sorting capability for modern controllers) implemented in some way directly in core architecture of `Grid`... btw. sorting on Category pages in context of two different stores doesn't work 🤷
3. As there's a decision to use `final`, I wasn't able to extend base `GridPositionUpdater`, I think that at this point we cannot implement any strategy class to handle this because it would introduce unacceptable breaking change?
4. I changed interface, and at this point I wonder if we should start with splitting between multi-store/single here, as this is breaking change

## TODO

- [ ] architectural decisions
- [x] sorting fully working in BO for single store context
- [x] sorting fully working in BO for single store context and different categories
- [x] refacto `CMS::updatePosition`
- [x] refacto `CMS::cleanPositions`
- [x] refacto `CMS::getLastPosition`
- [x] refacto `CMS::getCMSPages`
- [ ] check for backward compatibility issues
- [ ] document changes in `CMS` class
- [ ] check for tests implementation, eg. for `PositionUpdater`
- [ ] E2E

## DEPRECATED METHODS

- `CMS::listCms` - no `block_cms` in PrestaShop anymore, not possible to sort pages if they're are in few stores
- `CMS::getLinks` same as above, hard to get real CMS position, data between this and proper `CMS::getPages` are different

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/21352)

### CODE REVIEW

File: install-dev/data/db_structure.sql
Comment (PierreRambaud) [inline install-dev/data/db_structure.sql]: We can't delete SQL field like that, it's a huge BC, if people are using it, everything will fall in error :(

File: src/Core/Grid/Position/GridPositionMultiStoreUpdater.php
Comment (PierreRambaud) [inline src/Core/Grid/Position/GridPositionMultiStoreUpdater.php:83]: Is there a reason to create a method to call a PHP function alone :thinking: 
Look like a little bit overkill :sweat_smile:

File: src/Core/Grid/Position/PositionDefinition.php
Comment (PierreRambaud) [inline src/Core/Grid/Position/PositionDefinition.php]:
```suggestion
    public function getShopId(): ?int
```

File: src/Core/Grid/Position/PositionDefinitionInterface.php
Comment (PierreRambaud) [inline src/Core/Grid/Position/PositionDefinitionInterface.php]:
```suggestion
    public function getShopId(): ?int;
```

File: src/Core/Grid/Position/UpdateHandler/DoctrinePositionMultiStoreUpdateHandler.php
Comment (PierreRambaud) [inline src/Core/Grid/Position/UpdateHandler/DoctrinePositionMultiStoreUpdateHandler.php]:
```suggestion
        string $dbPrefix,
        int $shopId
```

File: src/Core/Grid/Position/UpdateHandler/DoctrinePositionMultiStoreUpdateHandler.php
Comment (PierreRambaud) [inline src/Core/Grid/Position/UpdateHandler/DoctrinePositionMultiStoreUpdateHandler.php:75]:
```suggestion
    public function getCurrentPositions(PositionDefinitionInterface $positionDefinition, int $parentId = null)
```

File: src/Core/Grid/Position/GridPositionMultiStoreUpdater.php
Comment (sowbiba) [inline src/Core/Grid/Position/GridPositionMultiStoreUpdater.php:27]:
```suggestion
declare(strict_types=1);

```

File: src/Core/Grid/Position/GridPositionMultiStoreUpdater.php
Comment (sowbiba) [inline src/Core/Grid/Position/GridPositionMultiStoreUpdater.php]:
```suggestion
    private function getNewPositions(PositionUpdateInterface $positionUpdate): array
```

File: src/Core/Grid/Position/UpdateHandler/DoctrinePositionMultiStoreUpdateHandler.php
Comment (sowbiba) [inline src/Core/Grid/Position/UpdateHandler/DoctrinePositionMultiStoreUpdateHandler.php:27]:
```suggestion
declare(strict_types=1);

```

**Review by sowbiba — COMMENTED** [review]
Some comments

File: src/Core/Grid/Position/GridPositionMultiStoreUpdater.php
Comment (kpodemski) [reply src/Core/Grid/Position/GridPositionMultiStoreUpdater.php:83]: Well, here you have a reason:

image

https://github.com/PrestaShop/PrestaShop/pull/10414

Maybe it's a bit more readable, I left it as in original file 😅 I don't know why it's not handled inside `getNewPositions`, if we want to get new positions anyway and in handler we don't really use position but iterator value 🤷‍♂️

**Comment by kpodemski** [comment]
@PrestaShop/prestashop-maintainers 

Hey y'all, I need:
- info if we can keep it as a GridPositionMultiStoreUpdater or we should switch to CQRS **or** leave it as is now and just add proper commands to use in different places?
- Grid tests are in `tests-legacy`, shall I add my tests there or move new set of tests for this implementation to `/tests`?
- can I create E2E tests or I should leave it to QA team?

**Comment by Julievrz** [comment]
Is there some wording in this PR? I can see the label but I wasn't able to find any wording inside the code... :thinking:

**Comment by kpodemski** [comment]
@Julievrz it seems that PrestonBot got lost because of some rebase 🤔

**Comment by Julievrz** [comment]
> @Julievrz it seems that PrestonBot got lost because of some rebase 🤔

Oh ok so if there is no wording I cancel the label. Thanks 😄

File: src/Core/Grid/Position/GridPositionMultiStoreUpdater.php
Comment (matthieu-rolland) [reply src/Core/Grid/Position/GridPositionMultiStoreUpdater.php:83]: That's one of the "clean code" precepts, a well-named function is better than having to write a comment explaining what is happening.

It could be considered overkill, but I kinda like it too...

**Review by matthieu-rolland — DISMISSED** [review]

File: src/Core/Grid/Position/GridPositionMultiStoreUpdater.php
Comment (PierreRambaud) [reply src/Core/Grid/Position/GridPositionMultiStoreUpdater.php:83]: It's overkill, you're not creating method each time you have to use `empty($array)` with name `checkIfArrayIsEmpty`.
I'm sad that a senior developer like him doesn't know the sort function which is a very old PHP4 function :(

Comment (matthieu-rolland) [reply src/Core/Grid/Position/GridPositionMultiStoreUpdater.php:83]: it's in between...  but you make a good point, the name of the variable `$newPositions` is self-explanatory, we could remove the comment and not use a function, either way, it is fine to me

Comment (matthieu-rolland) [reply src/Core/Grid/Position/GridPositionMultiStoreUpdater.php:83]: (this is the epitome of knit-picking though :D )

**Pull request closed without merging**
